kale/pipeline/multiomics_trainer.py

```diff
@@ -0,0 +1,232 @@
+
+        outputs = self.forward(x, adj_t, multi_modality=False)
+
+        for modality in range(self.num_modalities):
+            loss = self.loss_fn(outputs[modality], y[modality])
+            loss = torch.mean(torch.mul(loss, sample_weight[modality]))
+            self.logger.log_metrics({f"train_modality_step_loss ({modality + 1})": loss.detach()}, self.global_step)
+
+            optimizer[modality].zero_grad()
+            self.manual_backward(loss)
+            optimizer[modality].step()
+
+        if self.train_multi_modality_decoder and self.multi_modality_decoder is not None:
+            output = self.forward(x, adj_t, multi_modality=True)
+            multi_loss = self.loss_fn(output, y[0])
+            multi_loss = torch.mean(torch.mul(multi_loss, sample_weight[0]))
+            self.logger.log_metrics({"train_multi_modality_step_loss": multi_loss.detach()}, self.global_step)
+
+            optimizer[-1].zero_grad()
+            self.manual_backward(multi_loss)
+            optimizer[-1].step()
+
+    def test_step(self, test_batch, batch_idx: int):
+        """Compute and return the test loss.
+
+        Args:
+            test_batch (:class:`~torch.Tensor` | (:class:`~torch.Tensor`, ...) | [:class:`~torch.Tensor`, ...]):
+                The output of your :class:`~torch.utils.data.DataLoader`. A tensor, tuple or list.
+            batch_idx (int): Integer displaying index of this batch.
+        """
+        x = []
+        adj_t = []
+        y = []
+        for modality in range(self.num_modalities):
+            data = test_batch[modality]
+            x.append(data.x)
+            adj_t.append(data.adj_t)
+            y.append(torch.argmax(data.y[data.test_idx], dim=1))
+
+        if self.multi_modality_decoder is not None:
+            output = self.forward(x, adj_t, multi_modality=True)
+        else:
+            output = self.forward(x, adj_t, multi_modality=False)[0]
+
+        pred_test_data = torch.index_select(output, dim=0, index=test_batch[0].test_idx)
+        final_output = F.softmax(pred_test_data, dim=1).detach().cpu().numpy()
+        actual_output = y[0].detach().cpu()
+
+        if self.num_classes == 2:
+            self.log("Accuracy", round(accuracy_score(actual_output, final_output.argmax(1)), 3))
+            self.log("F1", round(f1_score(actual_output, final_output.argmax(1)), 3))
+            self.log("AUC", round(roc_auc_score(actual_output, final_output[:, 1]), 3))
+        else:
+            self.log("Accuracy", round(accuracy_score(actual_output, final_output.argmax(1)), 3))
+            self.log("F1 weighted", round(f1_score(actual_output, final_output.argmax(1), average="weighted"), 3))
+            self.log("F1 macro", round(f1_score(actual_output, final_output.argmax(1), average="macro"), 3))
+
+        return final_output
+
+    def _custom_data_loader(self) -> DataLoader:
+        """Return an iterable or collection of iterables specifying samples."""
```

Comment:
specifying is weird, specific? talk to clarify

Reply:
The same for below

Reply:
that specify all samples in the dataset.

Reply:
Changed.